### PR TITLE
Add erp bridge

### PR DESCRIPTION
**Pull request by schlenkibus**

### CODE REVIEW

File: build-environment/pi4-native/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+configure_file(${CMAKE_SOURCE_DIR}/package/c16-bridge/Cargo.toml .)
+configure_file(${CMAKE_SOURCE_DIR}/package/c16-bridge/Cargo.lock .)
+registerPod()
```
Comment (hhoegelo) [inline build-environment/pi4-native/CMakeLists.txt:3]: Pls use the DEPENDENCIES argument to make the pod dependent in Cargo.toml/lock

File: build-environment/pi4-native/Dockerfile.in [hunk 2]
```diff
@@ -0,0 +1,24 @@
+FROM docker.io/arm64v8/debian:bullseye-slim
+
+ENV LANG en_US.UTF-8
+ENV DEBIAN_FRONTEND=noninteractive
+
+RUN apt-get update -y && \
```
Comment (hhoegelo) [inline build-environment/pi4-native/Dockerfile.in:6]: do we really need this libraries for the rust build?

File: build-environment/pi4-native/readme.md [hunk 3]
```diff
@@ -0,0 +1,2 @@
+### TODO
```
Comment (hhoegelo) [inline build-environment/pi4-native/readme.md:1]: Is this still a todo?

File: package/audio-engine/src/AudioEngineOptions.cpp [hunk 4]
```diff
@@ -26,24 +26,24 @@ AudioEngineOptions::AudioEngineOptions(int &argc, char **&argv)
   Glib::OptionGroup mainGroup("common", "common options");
   Glib::OptionContext ctx;
```
Comment (hhoegelo) [inline package/audio-engine/src/AudioEngineOptions.cpp:27]: hmm this changes seem to be unrelated

File: package/audio-engine/src/main.cpp [hunk 5]
```diff
@@ -102,14 +102,16 @@ static std::unique_ptr<AudioOutput> createAudioOut(const std::string &name, Synt
```
Comment (hhoegelo) [inline package/audio-engine/src/main.cpp:102]: If the audio engine changes are indeed unrelated and it is not too much work, I'd appreciate  a separate PR for it.

File: package/c16-bridge/.cargo/config.toml [hunk 6]
```diff
@@ -0,0 +1,8 @@
+[target.armv7-unknown-linux-gnueabihf]
+linker = "arm-linux-gnueabihf-gcc"
+rustflags = [
+    "-C", "link-arg=--sysroot=/home/justus/rpi-sysroot"
```
Comment (hhoegelo) [inline package/c16-bridge/.cargo/config.toml:4]: this seems to be not portable

File: package/c16-bridge/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+set_target_properties(c16-bridge PROPERTIES LOCATION ${CMAKE_CURRENT_BINARY_DIR})
+
+add_test(NAME c16-bridge-test
+        COMMAND cargo test
+        WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
+
+install(FILES ${CMAKE_CURRENT_BINARY_DIR}/rootfs/bin/c16-bridge DESTINATION ${CMAKE_INSTALL_PREFIX}/bin)
+
+include(${CMAKE_BINARY_DIR}/cmake/package.cmake)
+deb("" "text fuer spaeter")
```
Comment (hhoegelo) [inline package/c16-bridge/CMakeLists.txt:29]: "text fuer spater" should be a short description of your package bundle

File: package/c16-bridge/libuart/uart.c [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+#include <fcntl.h>
+
+int transmit_uart(uint8_t const byte, int32_t const portFd)
+{
+  return write(portFd, &byte, sizeof byte);
+}
+
+int read_data(char* message_buffer, int32_t portFd)
+{
+  int bytesRead = read(portFd, message_buffer, 10);
```
Comment (hhoegelo) [inline package/c16-bridge/libuart/uart.c:21]: can we be sure that read does not block if there are less than 10 bytes available? If not, maybe we should configure the fd as NON_BLOCKING.

File: package/c16-bridge/libuart/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,4 @@
+cmake_minimum_required(VERSION 3.0)
```
Comment (hhoegelo) [inline package/c16-bridge/libuart/CMakeLists.txt:1]: Before merging, we should maybe put some effort in avoiding mixed language setup. I don't see why the uart stuff could not be written in rust, which would reduce complexity IMHO.

File: package/c16-bridge/src/application.rs [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+use serde::{Deserialize, Serialize};
+use std::io::{stdout, Write};
+use std::process::abort;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::mpsc::{channel, Receiver, Sender};
+use std::sync::{Arc, Mutex};
+use std::thread::JoinHandle;
+use ws::listen;
+
+#[allow(dead_code)]
```
Comment (hhoegelo) [inline package/c16-bridge/src/application.rs:19]: Why do we want to allow dead code?

File: package/playground/src/use-cases/LraUseCaseProxy.cpp [hunk 11]
```diff
@@ -0,0 +1,121 @@
+#include "LraUseCaseProxy.h"
```
Comment (hhoegelo) [inline package/playground/src/use-cases/LraUseCaseProxy.cpp:1]: not sure if LraUseCase should be merged, as we actually do not want to have this in product agnostic playground

File: package/c16-bridge/src/main.rs [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+pub(crate) mod uart_protocol;
+pub(crate) mod websocket_server;
+
+use crate::options::Args;
+
+fn main() {
+    let args = Args::parse();
+    let app = Application::new(format!("{}:{}", args.address, args.port), args.usb, args.fixed_update_target_rate);
+
+    std::thread::sleep(std::time::Duration::from_secs(1));
```
Comment (hhoegelo) [inline package/c16-bridge/src/main.rs:24]: if we really need to sleep here, pls let the code explain, why.

File: package/c16-bridge/src/bridge_funcs.rs [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+    const CHECK_SIZE: usize = 38;
+
+    let is_sysex = event.buffer[0] == 0xF0;
+    let has_correct_size = event.used_bytes == CHECK_SIZE;
+
+    if is_sysex && has_correct_size {
+        return Some(EventReceiverType::UIEvent);
+    }
+
+    Some(EventReceiverType::UIEvent)
```
Comment (hhoegelo) [inline package/c16-bridge/src/bridge_funcs.rs:19]: Maybe I am missing something, but this function doesn't seem to do much

File: package/c16-bridge/src/midi_sensor_sysex_parser.rs [hunk 14]
```diff
@@ -0,0 +1,324 @@
...
+    if result == TWENTYONE_ONES {
+        return result;
+    }
+
+    //HELLO HOW IS THIS POSSIBLLE?
+    //OG C code:
+    // if (result & 0b100000000000000000000)
+    //     result |= 0b11111111111000000000000000000000;
+    #[allow(clippy::bad_bit_mask)]
+    if result & 0b100000000000000000000 == 1 {
```
Comment (hhoegelo) [inline package/c16-bridge/src/midi_sensor_sysex_parser.rs:226]: from what I understand, this should be always false, no?

File: package/c16-bridge/src/midi_utils.rs [hunk 15]
```diff
@@ -0,0 +1,251 @@
...
+                .unwrap_or(SensorDataConfig { erp_smoothing: 0.1 }),
+            raw_erp_data: std::collections::HashMap::new(),
+            diff_erps: std::collections::HashMap::new(),
+            filtered_erp_data: std::collections::HashMap::new(),
+            raw_lsd_data: std::collections::HashMap::new(),
+            quantizer: std::collections::HashMap::new(),
+            last_angles: std::collections::HashMap::new(),
+        };
+
+        for i in 0..6 {
```
Comment (hhoegelo) [inline package/c16-bridge/src/midi_utils.rs:42]: Pls consider have a named constant for '6'

File: package/c16-bridge/src/midi_utils.rs [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+
+            let (tx, rx) = channel::<MidiEvent>();
+
+            match midi_in.connect(
+                port,
+                &name,
+                move |_stamp, midi_message, user_data: &mut MidiUserData| {
+                    match MidiEvent::new(create_midi_event_from_bytes(midi_message)) {
+                        Ok(midi_event) => {
+                            if let Some(event_type) = bridge_funcs::dispatch_from_midi(midi_event.clone()) {
```
Comment (hhoegelo) [inline package/c16-bridge/src/midi_utils.rs:209]: dispatch_from_midi should be happy with a const ref, so we can get rid of the clone

File: package/c16-bridge/src/uart_protocol.rs [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+const MAGIC2: u8 = ESC;
+const MAGIC3: u8 = STX;
+
+// typedef void (*UartReceiveComplete_Callback)(uint8_t contentId, uint8_t len, uint8_t *pData);
+// typedef int (*UartTransmit_Callback)(uint8_t data, uint32_t const lineStatus);
+
+//i32 -> fd
+#[allow(clippy::type_complexity)]
+pub struct TransmitProcessor<'a> {
+    callback: Box<dyn FnMut(&mut UartSender, u8, u32) -> bool + 'a>,
```
Comment (hhoegelo) [inline package/c16-bridge/src/uart_protocol.rs:28]: I am really not a rust expert, but I think you can get rid of all the life time annotations and boxing by defining callback as

```
callback: fn(&mut UartSender, u8, u32) -> bool,
```

**Comment by schlenkibus** [comment]
i also have 2 blocking issues that need resolving before merge:
- [ ] `make update-c16` does not include the c16-bridge binary
- [ ] when copying the c16-bridge.deb to the pi and installing it via `sudo dpkg -i` the resulting binary is not executable `chmod +x` fixes that

**Review by hhoegelo — APPROVED** [review]

File: package/c16-bridge/libuart/uart.c (hunk 8 above)
Comment (KlausStrohhaecker) [reply package/c16-bridge/libuart/uart.c:21]: > can we be sure that read does not block if there are less than 10 bytes available? If not, maybe we should configure the fd as NON_BLOCKING.

The whole idea is that the thread sleeps until read() returns something or a time out happens. See line 76 and 77:
  tty.c_cc[VTIME] = 1;  // blocking read with timeout...
  tty.c_cc[VMIN] = 0;   //...after 0.1sec

10 bytes is the worst-case length of messages sent from LPC to host.

Comment (KlausStrohhaecker) [reply package/c16-bridge/libuart/uart.c:21]: > do we need the ack from uart at all? If so, is the message really 10 bytes?
ACK Message can be 8 to 10 bytes because of "ESC"-doubling in the payload data (which is the TAN).
Yes we need th ACK, because that is the only way to know if a command sent to the LPC was accepted and processed properly.
Note that the returning TAN's may appear out or order. The sender has to keep track of what messages with what TANs were sent and if they have been acknowledged by the LPC after due time (say, within 300ms or so), otherwise they must be resent.

Also, ACK is very useful as a "is-LPC-alive?" indicator, besides the heartbeat (which is not implemented yet in the LPC code).
